# 4.9 - Intro to Data Visualization with Python - Task

## Part 1

### Step 3

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

# New installed libraries
import matplotlib.pyplot as plt
import seaborn as sns
import scipy

# Import dataframes
path = r"D:\0 - Data Analytics\4 - Python fundamentals for Data Analysts\Instacart Basket Analisys"

# Importing dataframes
ords_prods_merge = pd.read_pickle(
    os.path.join(path, "02 Data", "Prepared Data", "ords_prods_merge_V3.pkl")
)

# import dataframe customer from Original Data
df_customer = pd.read_csv(
    os.path.join(path, "02 Data", "Original Data", "customers.csv"), index_col=False
)


In [45]:
# check columns name of df_customer
df_customer

user_id First Name    Surnam  Gender           STATE  Age date_joined  \
0         26711    Deborah  Esquivel  Female        Missouri   48    1/1/2017   
1         33890   Patricia      Hart  Female      New Mexico   36    1/1/2017   
2         65803    Kenneth    Farley    Male           Idaho   35    1/1/2017   
3        125935   Michelle     Hicks  Female            Iowa   40    1/1/2017   
4        130797        Ann   Gilmore  Female        Maryland   26    1/1/2017   
...         ...        ...       ...     ...             ...  ...         ...   
206204   168073       Lisa      Case  Female  North Carolina   44    4/1/2020   
206205    49635     Jeremy   Robbins    Male          Hawaii   62    4/1/2020   
206206   135902      Doris  Richmond  Female        Missouri   66    4/1/2020   
206207    81095       Rose   Rollins  Female      California   27    4/1/2020   
206208    80148    Cynthia     Noble  Female        New York   55    4/1/2020   

        n_dependants fam_status  income  
0                  3    married  165665  
1                  0     single   59285  
2                  2    married   99568  
3                  0     single   42049  
4                  1    married   40374  
...              ...        ...     ...  
206204             1    married  148828  
206205             3    married  168639  
206206             2    married   53374  
206207             1    married   99799  
206208             1    married   57095  

[206209 rows x 10 columns]

In [46]:
# describe df_customer
df_customer.describe()

user_id            Age   n_dependants         income
count  206209.000000  206209.000000  206209.000000  206209.000000
mean   103105.000000      49.501646       1.499823   94632.852548
std     59527.555167      18.480962       1.118433   42473.786988
min         1.000000      18.000000       0.000000   25903.000000
25%     51553.000000      33.000000       0.000000   59874.000000
50%    103105.000000      49.000000       1.000000   93547.000000
75%    154657.000000      66.000000       3.000000  124244.000000
max    206209.000000      81.000000       3.000000  593901.000000

In [47]:
# shape of df_customer
df_customer.shape

(206209, 10)

In [48]:
# get the info of df_customer
df_customer.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206209 entries, 0 to 206208
Data columns (total 10 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   user_id       206209 non-null  int64 
 1   First Name    194950 non-null  object
 2   Surnam        206209 non-null  object
 3   Gender        206209 non-null  object
 4   STATE         206209 non-null  object
 5   Age           206209 non-null  int64 
 6   date_joined   206209 non-null  object
 7   n_dependants  206209 non-null  int64 
 8   fam_status    206209 non-null  object
 9   income        206209 non-null  int64 
dtypes: int64(4), object(6)
memory usage: 15.7+ MB


### 4
Wrangle the data so that it follows consistent logic; <br>
for example, rename columns with illogical names and drop columns that don’t add anything to your analysis

In [49]:
# first remane columns of df_customer so they are all lowercase, "Surname" to "last_name", fam_status to "marital_status", n_dependants to "number_of_dependants", income to "annual_income"
df_customer.rename(
    columns={
        "First Name": "first_name",
        "Surnam": "last_name",
        "Gender": "gender",
        "STATE": "state",
        "Age": "age",
        "fam_status": "marital_status",
        "n_dependants": "number_of_dependants",
        "income": "annual_income",
    },
    inplace=True,
)


In [50]:
# check columns name of df_customer
df_customer

user_id first_name last_name  gender           state  age date_joined  \
0         26711    Deborah  Esquivel  Female        Missouri   48    1/1/2017   
1         33890   Patricia      Hart  Female      New Mexico   36    1/1/2017   
2         65803    Kenneth    Farley    Male           Idaho   35    1/1/2017   
3        125935   Michelle     Hicks  Female            Iowa   40    1/1/2017   
4        130797        Ann   Gilmore  Female        Maryland   26    1/1/2017   
...         ...        ...       ...     ...             ...  ...         ...   
206204   168073       Lisa      Case  Female  North Carolina   44    4/1/2020   
206205    49635     Jeremy   Robbins    Male          Hawaii   62    4/1/2020   
206206   135902      Doris  Richmond  Female        Missouri   66    4/1/2020   
206207    81095       Rose   Rollins  Female      California   27    4/1/2020   
206208    80148    Cynthia     Noble  Female        New York   55    4/1/2020   

        number_of_dependants marital_status  annual_income  
0                          3        married         165665  
1                          0         single          59285  
2                          2        married          99568  
3                          0         single          42049  
4                          1        married          40374  
...                      ...            ...            ...  
206204                     1        married         148828  
206205                     3        married         168639  
206206                     2        married          53374  
206207                     1        married          99799  
206208                     1        married          57095  

[206209 rows x 10 columns]

We could maybe drop first_name and last_name for privacy protection, but for the moment i prerfer to keep it.

### 5
Complete the fundamental data quality and consistency checks you’ve learned throughout this Achievement; <br> 
for example, check for and address missing values and duplicates, and convert any mixed-type data.

### Missing values

In [51]:
# check for missing values
df_customer.isnull().sum()

user_id                     0
first_name              11259
last_name                   0
gender                      0
state                       0
age                         0
date_joined                 0
number_of_dependants        0
marital_status              0
annual_income               0
dtype: int64

We have 11259 missing values in the column first_name.<br> 
We don't need to address them casue they are not a key variable for the analysis, but this result might force us to drop the first_name and last_name columns.

### Duplicates

In [52]:
# check for duplicates
df_customer_dups = df_customer[df_customer.duplicated()]

df_customer_dups

Empty DataFrame
Columns: [user_id, first_name, last_name, gender, state, age, date_joined, number_of_dependants, marital_status, annual_income]
Index: []

It appears that there are not duplicates.

### Data consistency

In [53]:
# check the types of the columns\
df_customer.dtypes

user_id                  int64
first_name              object
last_name               object
gender                  object
state                   object
age                      int64
date_joined             object
number_of_dependants     int64
marital_status          object
annual_income            int64
dtype: object

In [55]:
# check for mixed types
# use function below to check for mixed types
for col in df_customer.columns.tolist():
    weird = (
        df_customer[[col]].applymap(type) != df_customer[[col]].iloc[0].apply(type)
    ).any(axis=1)
    if len(df_customer[weird]) > 0:
        print(col)

C:\Users\oppre\AppData\Local\Temp\ipykernel_14600\1452875470.py:5: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  df_customer[[col]].applymap(type) != df_customer[[col]].iloc[0].apply(type)
C:\Users\oppre\AppData\Local\Temp\ipykernel_14600\1452875470.py:5: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  df_customer[[col]].applymap(type) != df_customer[[col]].iloc[0].apply(type)
C:\Users\oppre\AppData\Local\Temp\ipykernel_14600\1452875470.py:5: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  df_customer[[col]].applymap(type) != df_customer[[col]].iloc[0].apply(type)
C:\Users\oppre\AppData\Local\Temp\ipykernel_14600\1452875470.py:5: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  df_customer[[col]].applymap(type) != df_customer[[col]].iloc[0].apply(type)
C:\Users\oppre\AppData\Local\Temp\ipykernel_14600\1452875470.py:5: FutureWarning: DataFrame.

first_name


C:\Users\oppre\AppData\Local\Temp\ipykernel_14600\1452875470.py:5: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  df_customer[[col]].applymap(type) != df_customer[[col]].iloc[0].apply(type)
C:\Users\oppre\AppData\Local\Temp\ipykernel_14600\1452875470.py:5: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  df_customer[[col]].applymap(type) != df_customer[[col]].iloc[0].apply(type)
C:\Users\oppre\AppData\Local\Temp\ipykernel_14600\1452875470.py:5: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  df_customer[[col]].applymap(type) != df_customer[[col]].iloc[0].apply(type)
C:\Users\oppre\AppData\Local\Temp\ipykernel_14600\1452875470.py:5: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  df_customer[[col]].applymap(type) != df_customer[[col]].iloc[0].apply(type)
C:\Users\oppre\AppData\Local\Temp\ipykernel_14600\1452875470.py:5: FutureWarning: DataFrame.

The column first_name has some different datatypes, this may be caused by the missing values. <br>
We will change it to string, but we might not aggreagate it later.

In [56]:
# change the type of the columns
df_customer["first_name"] = df_customer["first_name"].astype("str")

In [57]:
# check again for mixed types
for col in df_customer.columns.tolist():
    weird = (
        df_customer[[col]].applymap(type) != df_customer[[col]].iloc[0].apply(type)
    ).any(axis=1)
    if len(df_customer[weird]) > 0:
        print(col)

C:\Users\oppre\AppData\Local\Temp\ipykernel_14600\1156895741.py:4: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  df_customer[[col]].applymap(type) != df_customer[[col]].iloc[0].apply(type)
C:\Users\oppre\AppData\Local\Temp\ipykernel_14600\1156895741.py:4: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  df_customer[[col]].applymap(type) != df_customer[[col]].iloc[0].apply(type)
C:\Users\oppre\AppData\Local\Temp\ipykernel_14600\1156895741.py:4: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  df_customer[[col]].applymap(type) != df_customer[[col]].iloc[0].apply(type)
C:\Users\oppre\AppData\Local\Temp\ipykernel_14600\1156895741.py:4: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  df_customer[[col]].applymap(type) != df_customer[[col]].iloc[0].apply(type)
C:\Users\oppre\AppData\Local\Temp\ipykernel_14600\1156895741.py:4: FutureWarning: DataFrame.

Data types looks consistent trhough the columns.

## 6
Combine your customer data with the rest of your prepared Instacart data. <br>
(Hint: Make sure the key columns are the same data type!)

In [58]:
# check data types of the columns of prods_ords_merge
ords_prods_merge.dtypes

order_id                    int64
user_id                     int64
order_number                int64
orders_day_of_week          int64
order_hour_of_day           int64
days_since_last_order     float64
first_order                  bool
product_id                  int64
add_to_cart_order           int64
reordered                   int64
product_name               object
aisle_id                    int64
department_id               int64
prices                    float64
_merge                   category
price_label                object
busiest_day                object
busiest_days               object
busiest_period_of_day      object
max_order                   int64
loyalty_flag               object
avg_product_price         float64
spending_flag              object
median_between_orders     float64
order_frequency_flag       object
dtype: object

user_id is the common key and tehy have the same datatype. <br>
We will need though to drop the "_merge" column to proceed with the merge.

In [59]:
# drop _merge column
ords_prods_merge.drop(columns=["_merge"], inplace=True)

In [60]:
# merge the dataframes
ords_prods_customers_merged = ords_prods_merge.merge(
    df_customer, on="user_id", indicator=True
)

In [61]:
# check the merge
ords_prods_customers_merged["_merge"].value_counts()

_merge
both          32404859
left_only            0
right_only           0
Name: count, dtype: int64

Merged successfully.

## 8
Export this new dataframe as a pickle file so you can continue to use it in the second part of this task.

In [62]:
# export the merged dataframe
ords_prods_customers_merged.to_pickle(
    os.path.join(path, "02 Data", "Prepared Data", "ords_prods_customers_merged_V1.pkl")
)